src/Chapter3.hs

```diff
@@ -1006,6 +1139,64 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+class Contestant a where
+  fightIsOver :: a -> Bool -- fight is over if the Contestant is either dead or decided to run away. This results in a win for the other contestant
+  handleAttack :: a -> Attack -> a -- the other contestant attacks this contestant with a given Attack
+  applyNextAction :: (Contestant b) => a -> b -> (a, b)
+
+data Knight = Knight {
+  knightHealth :: Health,
+  knightDefense :: Defense,
+  knightActions :: [KnightAction]
+} deriving Show
+
+data KnightAction = AttackContestant Attack | DrinkAHealthPotion Health | CastASpell Defense deriving Show
+
+instance Contestant Knight where
+  fightIsOver knight = knightHealth knight <= Health 0
+  handleAttack knight attack = knight {knightHealth = calculateNewKnightHealth knight attack}
+  applyNextAction k@(Knight _ _ []) c = (k, c)
+  applyNextAction k@(Knight _ _  actions@(AttackContestant attack : _)) b = (k { knightActions = rotate actions}, handleAttack b attack)
+  applyNextAction k@(Knight health _ actions@(DrinkAHealthPotion moreHealth : _)) b = (k { knightHealth = health + moreHealth, knightActions = rotate actions}, b)
+  applyNextAction k@(Knight _ defense actions@(CastASpell moreDefense : _)) b = (k { knightDefense = defense + moreDefense, knightActions = rotate actions}, b)
```

Comment:
I think this implementation of duplicating main function name and pattern is a bit difficult to comprehend. I would recommend using `case-of` by pattern-matching on the list (empty or not) and then another `case-of` to pattern-match on the first action.

Reply:
Yes, that's way better that way!